### DISCUSSION TITLE
reduced "checkpoint" for inference only

### DISCUSSION THREAD

**Discussion by wunderbarb** [gomlx/gomlx#408] [General]
# Inference Only
## Objective
Currently, the checkpoints contain everything ranging from model weights to the global step training data.  Inference needs only a subpart of this data.
The aim is to create an inference data set that holds only the necessary information.
The following constraints are defined:

- The inference data is derived from checkpoint data.
- The inference may load inference date or checkpoint seamlessly.
- This modification must be fully compatible with the current training process.

## Foreseen modifications
Add a new method to `Handler` named `SaveForInference(path string) error` that takes from an already loaded checkpoint the necessary data.  The new data are stored in the same format as the current checkpoint.
### What gets excluded
Variables excluded (identified by scope/name from `ParameterName`):
Scope starts with `/AdamOptimizer` (`optimizers.AdamDefaultScope`) — Adam 1st/2nd moments
Scope starts with `/optimizers` (`optimizers.Scope`) — learning rate variable, cosine schedule state
Name is `global_step` in root scope (`optimizers.GlobalStepVariableName`)
Name is `#rngState` in root scope (`context.RNGStateVariableName`)
Params excluded (by key):
`learning_rate`, `optimizer`, `clip_step_by_value`, `clip_nan`, `nanlogger`
`adam_epsilon`, `adam_dtype`, `adam_weight_decay`, `adam_beta1`, `adam_beta2`, `adam_backoff`

- Update `gomlx_checkpoints` to add an export for inference command
- Update `mnist` example to export an inference checkpoint at the end of training and update the `classifier`
### Questions

- What information that can be stripped did I miss?
- Should we propose a naming for the inference data only?
- Should we create a function `LoadForInference` to discriminate?  In that case, the `Load` could verify that the training information is present.

**Comment by janpfeifer** [comment]
Indeed, the full saving for inference is not fully developed, thanks for bringing it up. 

Some is in place already though. Let me try to add some thoughts/context:

I imagine users may have 3 different reasons (?) to want to trim-down models for inference only:

1. Save memory (GPU) and loading time.
2. Save disk space, specially for distributing pre-trained model.
3. Hide training details -- if it's proprietary knowledge.

 Current tools address (somewhat) (1) and (2):
 
 1. `checkpoints` have a "lazy" loading of variables: it only prefills the variable existance, but only actually loads variables once it's used. So using a checkpoint for inference won't incur in loading training data (no GPU/Host memory use, no loading time) -- except if it's broken and I was not aware.
 2. The `cmd/gomlx_checkpoint -delete_var` to delete the variables under a scope. Remove `/AdamOptimizer` and whatever training variable space left if negligible. The issue here is, it requires one to know what that scope is. Maybe we should add a `-trim_for_inference` flag ? 

Now for (3) **hiding training details**: doing it automatically would be more work, because it would involve somehow tagging each hyper-parameter if it is training/inference only, and it can only be done case by case (each layer library would have to tag its own hyper-parameters)... Let me know if this is something you need, I would otherwise leave it for later, as lower priority for now.

Another aspect of your suggestion is adding a `SaveForInference`, as opposed to "trim model once we are happy training with it". I always favored the later, because during training I'm not sure when I'm happy training with it. I often train, then evaluate it in some validation data ... and only after selecting the best of many models trained, would I trim the winning model and distribute it. What do you imagine would be the use case for a `SaveForInference` method ?

What do you think ?
